FreeRTOS-Plus/Demo/coreMQTT_Windows_Simulator/MQTT_Multitask/DemoTasks/simple_sub_pub_demo.c

```diff
@@ -0,0 +1,546 @@
+    xCommandParams.pCmdCompleteCallbackContext = &xCommandContext;
+
+    /* For a finite number of publishes... */
+    for( ulValueToNotify = 0UL; ulValueToNotify < mqttexamplePUBLISH_COUNT; ulValueToNotify++ )
+    {
+        /* Create a payload to send with the publish message.  This contains
+         * the task name and an incrementing number. */
+        snprintf( payloadBuf,
+                  mqttexampleSTRING_BUFFER_LENGTH,
+                  "%s publishing message %d",
+                  taskName,
+                  ( int ) ulValueToNotify );
+
+        xPublishInfo.payloadLength = ( uint16_t ) strlen( payloadBuf );
+
+        /* Also store the incrementing number in the command context so it can
+         * be accessed by the callback that executes when the publish operation
+         * is acknowledged. */
+        xCommandContext.ulNotificationValue = ulValueToNotify;
+
+        LogInfo( ( "Sending publish request to agent with message \"%s\" on topic \"%s\"",
+                   payloadBuf,
+                   pcTopicBuffer ) );
+
+        /* To ensure ulNotification doesn't accidentally hold the expected value
+         * as it is to be checked against the value sent from the callback.. */
+        ulNotification = ~ulValueToNotify;
+
+        xCommandAdded = MQTTAgent_Publish( &xGlobalMqttAgentContext,
+                                           &xPublishInfo,
+                                           &xCommandParams );
+        configASSERT( xCommandAdded == MQTTSuccess );
+
+        /* For QoS 1 and 2, wait for the publish acknowledgment.  For QoS0,
+         * wait for the publish to be sent. */
+        LogInfo( ( "Task %s waiting for publish %d to complete.",
+                   taskName,
+                   ulValueToNotify ) );
+        prvWaitForCommandAcknowledgment( &ulNotification );
+
+        /* The value received by the callback that executed when the publish was
+         * acked came from the context passed into MQTTAgent_Publish() above, so
+         * should match the value set in the context above.  However QoS 0 does
+         * not provide guaranteed delivery so it is ok for the values not to match
+         * if xQos is 0.*/
+        configASSERT( ( ulNotification == ulValueToNotify ) || ( xQoS == 0 ) );
+
+        if( ulNotification == ulValueToNotify )
+        {
+            LogInfo( ( "Received ack from publishing to topic %s.",
+                       pcTopicBuffer ) );
+        }
+        else
+        {
+            if( xQoS == 0 )
+            {
+                ulQoS0FailureCount++;
+            }
+
+            LogInfo( ( "Error - Timed out or didn't receive ack from publishing to topic %s",
+                       pcTopicBuffer ) );
+        }
+
+        /* Add a little randomness into the delay so the tasks don't remain
+         * in lockstep. */
+        xTicksToDelay = pdMS_TO_TICKS( mqttexampleDELAY_BETWEEN_PUBLISH_OPERATIONS_MS ) +
+                        ( uxRand() % 0xff );
```

Comment:
Should corePKCS11 be used to generate a random number instead of using `uxRand()`? AFAIK, `uxRand` represents a legacy PRNG function which we should stray away from to suggest users to use corePKCS11 for possible use of TRNG

Reply:
The mutual auth demo uses `uxRand()`, I don't think it's necessary to add pkcs11 to this demo since it's not what it's demonstrating. If there's an improvement to switch from using `uxRand()` to pkcs11, it should be in another PR, and affect all the demos

Reply:
Yeah this delay is just to make the demo pretty, it is not cryptogaphically important.